cli/azd/pkg/infra/provisioning/bicep/bicep_cache.go

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package bicep
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	"github.com/azure/azure-dev/cli/azd/pkg/azure"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment/azdcontext"
+	"github.com/azure/azure-dev/cli/azd/pkg/lazy"
+	"github.com/azure/azure-dev/cli/azd/pkg/osutil"
+	"golang.org/x/exp/slices"
+)
+
+// cache is serialized into json and represents the local provisioning cache for bicep.
+type cache struct {
+	Template   json.RawMessage
+	Parameters azure.ArmParameters
+}
+
+// CacheManager provides the functionality for using bicep cache.
+type CacheManager interface {
+	// Current returns existing cache or nil when there's no cache.
+	Current(context context.Context) *cache
+	// Cache persist cache. Use Cache(nil) to clear cache.
+	Cache(context context.Context, cache *cache) error
+	// Equal compares cache against current and return false when they are different.
+	Equal(context context.Context, cache *cache) bool
+}
+
+// bicepCache writes a cache file inside the current azd environment.
+// Remote cache is also supported by setting the env var:
+//
+// AZURE_BICEP_CACHE_CONFIG=type,args...
+//
+// where supported `type` can be:
+// - azureBlob
+//
+// `args` depends on the type as follows:
+//
+// **azureBlob args: `containerName,connectionString`,
+// example:
+//
+//	AZURE_BICEP_CACHE_CONFIG=azureBlob,cache,DefaultEndpointsProtocol=https;Ac....
+//
+// If the manager fails to parse or connect to the remote config, it will fallback automatically to use local file cache.
+type cacheClient struct {
+	lazyAzdContext *lazy.Lazy[*azdcontext.AzdContext]
+	lazyAzdEnv     *lazy.Lazy[*environment.Environment]
+	// for testing purposes, allows to mock the implementation for reading/writing cache.
+	overrideReadFunc  readCache
+	overrideWriteFunc writeCache
+}
+
+const c_cacheFileName = "bicep.cache"
+
+type readCache func(context context.Context, arg any) ([]byte, error)
+type writeCache func(context context.Context, arg any, cache []byte) error
+
+func cacheFromAzureBlob(context context.Context, arg any) ([]byte, error) {
+	azBlobData, goodCast := arg.(*azBlobSource)
+	if !goodCast {
+		return nil, fmt.Errorf("unexpected data for pulling cache from Azure Blob.")
+	}
+	client, err := azblob.NewClientFromConnectionString(azBlobData.azStorageConnectionString, nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating client from connection string: %w", err)
+	}
+
+	blobDownloadResponse, err := client.DownloadStream(context, azBlobData.azContainerName, c_cacheFileName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("downloading blob from azure: %w", err)
+	}
+	return io.ReadAll(blobDownloadResponse.Body)
+}
```

Comment:
Lets discuss how we could leverage the same azure blob storage client that I am using in remote state for this as well.  Take a look at the Remote env PR and lets think to see if we can leverage a similar concept of remote data stores that can have their own implementations rather than hard coding multiple implementations into the cache client.